### PR TITLE
s:dispatch: shorten displayed message to avoid hit-enter prompt

### PR DESCRIPTION
**Pull request by blueyed**
Fixes https://github.com/tpope/vim-dispatch/issues/166.

### CODE REVIEW

**Comment by blueyed** [comment]
See https://github.com/vim-scripts/ingo-library/blob/558132e2221db3af26dc2f2c6756d092d48a459f/autoload/ingo/avoidprompt.vim#L31 for a more evolved method to get the maxlength correctly (instead of just `-12`).

File: autoload/dispatch.vim [hunk 1]
```diff
@@ -215,8 +215,21 @@ function! s:dispatch(request) abort
...
       let pid = dispatch#pid(a:request)
-      echo ':!'.a:request.expanded .
-            \ ' ('.handler.'/'.(!empty(pid) ? pid : '?').')'
+
+      " Display cmd with handler and pid, avoiding a hit-enter prompt.
+      let msg = ':!'
+      let suffix = ' ('.handler.'/'.(!empty(pid) ? pid : '?').')'
+      let cmd = a:request.expanded
+      " NOTE: the extra "-12" is required to avoid the hit-enter, although
+      " it's displayed on a single line already?!
```
Comment (blueyed) [inline autoload/dispatch.vim]: More evolved: https://github.com/vim-scripts/ingo-library/blob/558132e2221db3af26dc2f2c6756d092d48a459f/autoload/ingo/avoidprompt.vim#L31

**Comment by blueyed** [comment]
This should also be done in `dispatch#complete` then.

**Comment by blueyed** [comment]
What about the suggestion from https://github.com/tpope/vim-dispatch/issues/166#issuecomment-201405421:

> A shorter/simpler approach would be to just call `:redraw` after the `:echom` (instead of `:echo`).

This would display it normally, but skip it in case there would have been a hit-enter prompt.
The benefit here would be that it's in `:mess`, in case you want to get to it later.

**Comment by petobens** [comment]
Hi @tpope and @blueyed; I just bumped into this issue and was wondering if there is a chance to move this pr forward (with the corresponding necessary changes). Thanks in advance!

**Comment by simlei** [comment]
Would love to see this PR or a similar change in vim-dispatch.

**Comment by blueyed** [comment]
Rebased (old head: 0722f74).

File: autoload/dispatch.vim [hunk 2]
```diff
@@ -373,7 +373,19 @@ function! s:dispatch(request) abort
...
       let a:request.handler = handler
       redraw
-      echo ':!'.a:request.expanded s:postfix(a:request)
+      let msg = ':!'
+      let suffix = s:postfix(a:request)
+      let cmd = a:request.expanded
+      " NOTE: the extra "-13" is required to avoid the hit-enter, although
+      " it's displayed on a single line already?!
+      let max_cmd_len = (&cmdheight * &columns)-2-len(suffix)-13
+      if len(cmd) > max_cmd_len
```
Comment (tpope) [inline autoload/dispatch.vim]: `len()` is byte length, it only works for ASCII. Use `len(substitute(cmd, '.', '.', 'g'))` to get the number of characters.

Maybe this explains the 12/13 discrepancy?

File: autoload/dispatch.vim [hunk 3]
```diff
@@ -373,7 +373,19 @@ function! s:dispatch(request) abort
...
+        let msg .= cmd[0:max_cmd_len-2] . '…'
```
Comment (tpope) [inline autoload/dispatch.vim]: Use `'<' . matchstr(cmd, '\{' . max_cmd_length - 1 . '\}$')` to get the correct length and match Vim's usual truncation (avoiding any encoding issues from the ellipsis in the process).

**Comment by blueyed** [comment]
> Maybe this explains the 12/13 discrepancy?

No.
But with Vim I only need `-2` (which makes more sense)!  (I was using Neovim apparently already when reporting the issue in 2016)
This is the case with Vim 7.4.2200 already (the oldest to really support async (#283)).
But have checked it using `echo repeat('X', &columns-1)` back to Vim 7.4.41.

Requiring more there is a bug in Neovim, and this patch here should maybe cope for that, but I'll look into getting it fixed there also.

**Comment by blueyed** [comment]
The difference is due to (not) using `set showcmd` etc.
Something like this would be useful: https://github.com/Shougo/unite.vim/blob/e1d3bb9f836a8645807b790345b69666bf22bb2f/autoload/unite/view.vim#L799-L826

**Comment by blueyed** [comment]
Even with (re)setting showcmd it will overwrite the tail when (re)setting it.
I've filed an issue for Vim itself: https://github.com/neovim/neovim/issues/10601.

But until it is fixed (and for older versions) I think we should check the two settings (ruler, showcmd) and take those into account for the max length.

File: autoload/dispatch.vim [hunk 4]
```diff
@@ -375,12 +375,14 @@ function! s:dispatch(request) abort
...
+      let suffix_len = len(substitute(suffix, '.', '.', 'g'))
       let cmd = a:request.expanded
+      let cmd_len = len(substitute(cmd, '.', '.', 'g'))
       " NOTE: the extra "-13" is required to avoid the hit-enter, although
       " it's displayed on a single line already?!
       let max_cmd_len = (&cmdheight * &columns)-2-len(suffix)-13
-      if len(cmd) > max_cmd_len
-        let msg .= cmd[0:max_cmd_len-2] . '…'
+      if cmd_len > max_cmd_len
+        let msg .= '<' . matchstr(cmd, '\v.{'.(max_cmd_len - 1).'}$')
```
Comment (blueyed) [inline autoload/dispatch.vim]: Initially this would display the beginning, now it displays then end/tail.

Comment (tpope) [reply autoload/dispatch.vim]: I think that's fine, the most important part is this `[whatever/123]` at the end anyways.

**Comment by blueyed** [comment]
For reference: `v:echospace` was added in Vim patch 8.1.1913 (https://github.com/vim/vim/commit/37f4cbd46f5a6f2dd3a48d5fa4324dce37e4bd6c).

**Comment by petobens** [comment]
@blueyed will that allow to close https://github.com/tpope/vim-dispatch/issues/166?

**Comment by blueyed** [comment]
@petobens 
Yes, it's meant to help with that.

**Pull request merged**